linter/linter.go

```diff
@@ -75,23 +79,42 @@ func New(realIssuer *x509.Certificate, realSigner crypto.Signer, skipLints []str
 	if err != nil {
 		return nil, err
 	}
-	return &Linter{lintIssuer, lintSigner, reg}, nil
+	return &Linter{lintIssuer, lintSigner, reg, realSigner.Public()}, nil
 }
 
 // Check signs the given TBS certificate using the Linter's fake issuer cert and
 // private key, then runs the resulting certificate through all non-filtered
-// lints. It returns an error if any lint fails. On success it also returns the
-// DER bytes of the linting certificate.
+// lints. If the subjectPubKey is identical to the public key of the real signer
+// used to create this linter, then the throwaway cert will have its pubkey
+// replaced with the linter's pubkey so that it appears self-signed. It returns
+// an error if any lint fails. On success it also returns the DER bytes of the
+// linting certificate.
 func (l Linter) Check(tbs *x509.Certificate, subjectPubKey crypto.PublicKey) ([]byte, error) {
-	lintCertBytes, cert, err := makeLintCert(tbs, subjectPubKey, l.issuer, l.signer)
+	lintPubKey := subjectPubKey
```

Comment:
This looked very similar to https://github.com/letsencrypt/boulder/pull/6996/files#diff-694bc9a02352f68c4a0beb8e111ddb9932c49bbaef81e88a0517c76ceb61f6c7 and I rewrote it locally after rebasing ontop of that PR.

What do you think about this instead?
```
func (l Linter) Check(tbs *x509.Certificate, subjectPubKey crypto.PublicKey) ([]byte, error) {
	lintPubKey := subjectPubKey
	keysEqual, err := core.PublicKeysEqual(l.realPubKey, lintPubKey)
	if err != nil {
		return nil, err
	}
	if keysEqual {
		lintPubKey = l.signer.Public()
	}

	lintCertBytes, cert, err := makeLintCert(tbs, lintPubKey, l.issuer, l.signer)
	if err != nil {
		return nil, err
	}

	lintRes := zlint.LintCertificateEx(cert, l.registry)
	err = ProcessResultSet(lintRes)
	if err != nil {
		return nil, err
	}

	return lintCertBytes, nil
}
```

Reply:
Yes, that PR was inspired by this one (see Jacob's comments above). But since it looks like that one will land first, I'm happy to update this one to use it.